### PR TITLE
[language, move] Introduce public key validation

### PR DESCRIPTION
**Pull request by huitseeker**
TL;DR: Introduce a native function for public key validation, and use it in TR key rotation.

This solves the problem that any 32-byte value would be accepted as a public key in the stdlib contracts, despite the length check being insufficient to protect against [invalid point attacks, small subgroup attacks](https://link.springer.com/content/pdf/10.1007/3-540-44598-6_8.pdf), and typos. 

In the last case, before this PR, a user would typically rotate their key to an invalid public key unwittingly, and then fail to produce signatures against that new key, ~~losing the balance of their account in the process~~ losing the ability to demonstrate the legality of their transactions. We'd rather have early failure.

This introduces a native function for checking the validity of this public key, and uses it upon rotation. It would probably be useful to call this public key validation function in a few more places - in fact, anywhere externally-controlled key material is registered on-chain. I have not looked beyond the key rotation contracts — Edit : and publication — for now.

A further step would be an opt-in proof of ownership (TBD).

### CODE REVIEW

File: language/move-lang/tests/functional/shared_ed25519_public_key/shared_key.move [hunk 1]
```diff
@@ -61,3 +64,19 @@ fun main() {
 }
 // check: ABORTED
 // check: 7000
+
+// rotating to a key with a good length but bad contents should fail
```
Comment (valerini) [inline language/move-lang/tests/functional/shared_ed25519_public_key/shared_key.move:68]: Do we want to have separate tests for failure cases:
1) point not on curve,
2) point of small order,
3) point not canonically serialized?

File: language/stdlib/modules/approved_payment.move [hunk 2]
```diff
@@ -65,7 +65,15 @@ module ApprovedPayment {
 
     // Rotate the key used to sign approved payments. This will invalidate any approved payments
     // that are currently in flight
-    public fun rotate_key(approved_payment: &mut T, new_public_key: vector<u8>) {
+    public fun rotate_key(approved_payment: &mut T, new_public_key: vector<u8>)
+    {
```
Comment (valerini) [inline language/stdlib/modules/approved_payment.move:69]: Might want to return { to the previous line.

File: language/move-lang/tests/functional/shared_ed25519_public_key/shared_key.move (hunk 1 above)
Comment (huitseeker) [reply language/move-lang/tests/functional/shared_ed25519_public_key/shared_key.move:68]: Not for now, as the native function does not expose them.
 It may be an improvement if Move has the error code breadth.

File: language/move-prover/tests/sources/stdlib/modules/signature.move [hunk 3]
```diff
@@ -1,6 +1,7 @@
 address 0x0 {
 
 module Signature {
+    native public fun ed25519_validate_pubkey(public_key: vector<u8>): bool;
```
Comment (sblackshear) [inline language/move-prover/tests/sources/stdlib/modules/signature.move:4]: Nit: comment would be good here (I know it's missing for the others, boo).

File: language/stdlib/modules/shared_ed25519_public_key.move [hunk 4]
```diff
@@ -34,6 +35,11 @@ module SharedEd25519PublicKey {
 
     fun rotate_key(shared_key: &mut T, new_public_key: vector<u8>) {
         Transaction::assert(Vector::length(&new_public_key) == 32, 7000);
```
Comment (sblackshear) [inline language/stdlib/modules/shared_ed25519_public_key.move:37]: Can probably kill this check now (since `validate_pubkey` presumably returns false for a bad length)?

File: language/move-lang/tests/functional/shared_ed25519_public_key/shared_key.move (hunk 1 above)
Comment (sblackshear) [reply language/move-lang/tests/functional/shared_ed25519_public_key/shared_key.move:68]: I think it would be useful to have these tests + maybe a few others (e.g. key too long, key too short) in `functional/natives/signature.move`. I don't think we want/need a more sophisticated error codes to write these tests--checking that the function says `false` for all of them is enough.

**Comment by junkil-park** [comment]
I confirm that the latest commit (f1aeaab) (with modifying `prelude.bpl`) passed the Prover test in CI (https://github.com/mimoo/mirai-bot/runs/668772188?check_suite_focus=true#step:7:1). Miraibot is silent when it finds no errors.

**Comment by davidiw** [comment]
Thanks for tackling this important endeavor!

File: language/stdlib/modules/shared_ed25519_public_key.move (hunk 4 above)
Comment (huitseeker) [reply language/stdlib/modules/shared_ed25519_public_key.move:37]: The nice thing you have with this is the consistent error code for a length error.

File: language/move-lang/tests/functional/shared_ed25519_public_key/shared_key.move (hunk 1 above)
Comment (huitseeker) [reply language/move-lang/tests/functional/shared_ed25519_public_key/shared_key.move:82]: Done

File: language/move-lang/tests/functional/natives/signature.move [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    // concatenation of the two above
+    let long_pubkey = x"1003d4017c3e843895a92b70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4660c";
+    let invalid_pubkey = x"0000000000000000000000000000000000000000000000000000000000000000";
+
+    Transaction::assert(Signature::ed25519_validate_pubkey(valid_pubkey) == true, 9003);
+    Transaction::assert(Signature::ed25519_validate_pubkey(short_pubkey) == false, 9003);
+    Transaction::assert(Signature::ed25519_validate_pubkey(long_pubkey) == false, 9003);
+    Transaction::assert(Signature::ed25519_validate_pubkey(invalid_pubkey) == false, 9003);
+}
+}
```
Comment (sblackshear) [inline language/move-lang/tests/functional/natives/signature.move:21]: <3

File: language/move-lang/tests/functional/natives/signature.move [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+fun main() {
+
+    // from RFC 8032
+    let valid_pubkey =  x"3d4017c3e843895a92b70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4660c";
+    let short_pubkey = x"100";
+    // concatenation of the two above
+    let long_pubkey = x"1003d4017c3e843895a92b70aa74d1b7ebc9c982ccf2ec4968cc0cd55f12af4660c";
+    let invalid_pubkey = x"0000000000000000000000000000000000000000000000000000000000000000";
+
+    Transaction::assert(Signature::ed25519_validate_pubkey(valid_pubkey) == true, 9003);
```
Comment (sblackshear) [inline language/move-lang/tests/functional/natives/signature.move:16]: Nit: no need for `== true` (and can do `!` for `== false`).
